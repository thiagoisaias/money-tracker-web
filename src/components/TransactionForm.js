import React, { Component } from "react";
import styled from "styled-components";
import moment from "moment";
import axios from "axios";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";

const FormContainer = styled.form`
  background-color: #fff;
  box-shadow: 0.5px 1px 1px 1px #ddd;


  > h2 {
    font-size: 1.5em;
    margin-bottom: 32px;
  }

  @media (max-width: 736px) {
    padding: 16px;
  }

  @media (min-width: 737px) {
    max-width: 900px;
    padding: 32px;
    margin: 32px auto;
    border-radius: 2px;
  }
`;

const TransactionDatePicker = styled(DatePicker)`
  display: block;
  height: 30px;
  width: 150px;
  padding-left: 16px;
  border: 1px solid #e2e2e2;
  border-radius: 2px;
  margin: 8px 0;
  font-size: 14px;
  cursor: pointer;
`;

const FormGroup = styled.div`
  margin: 16px 0;
`;

const FormLabel = styled.label`
  display: block;
`;

const FormInput = styled.input`
  display: block;
  height: 30px;
  padding-left: 16px;
  border: 1px solid #e2e2e2;
  border-radius: 2px;
  margin: 8px 0;
  font-size: 14px;
  width: ${props => (props.bigger ? "300px" : null)};
`;

const FormSelect = styled.select`
  display: block;
  margin: 8px 0;
`;

const SubmitButton = styled.button`
  width: 125px;
  height: 40px;
  background-color: #8378f4;
  cursor: pointer;
  border-radius: 2px;
  color: #f2f2f2;
  font-size: 12px;

  &:hover {
    opacity: 0.9;
  }

  @media (max-width: 736px) {
    margin: 32px auto;
  }

  @media (min-width: 737px) {
    margin-top: 55px;
  }
`;

class TransactionForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      categoriesList: [],
      formData: {
        description: "",
        value: "",
        category: 0,
        transactionType: 0,
        date: moment()
      }
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleDateChange = this.handleDateChange.bind(this);
  }

  componentDidMount() {
    axios
      .get("/categories")
      .then(response => {
        this.setState({
          categoriesList: response.data
        });
      })
      .catch(error => {
        console.log(error);
      });
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      formData: {
        ...this.state.formData,
        [name]: value
      }
    });
  }

  handleDateChange(event) {
    const selectedDate = event;
    this.setState({
      formData: {
        ...this.state.formData,
        date: selectedDate
      }
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    console.log("Send form data to API", this.state.formData);
    // this.props.history.replace("/");
  }

  render() {
    const options = this.state.categoriesList.map(category => {
      return (
        <option key={category.id} value={category.id}>
          {category.name}
        </option>
      );
    });

    return (
      <FormContainer>
        <h2>Transaction</h2>
        <FormGroup>
          <FormLabel>Transaction Type</FormLabel>
          <FormSelect name="transactionType" onChange={this.handleInputChange}>
            <option value="0">Expense</option>
            <option value="1">Earning</option>
          </FormSelect>
        </FormGroup>
        <FormGroup>
          <FormLabel>Description</FormLabel>
          <FormInput
            bigger
            name="description"
            type="text"
            onChange={this.handleInputChange}
          />
        </FormGroup>
        <FormGroup>
          <FormLabel>Value</FormLabel>
          <FormInput
            name="value"
            type="number"
            onChange={this.handleInputChange}
          />
        </FormGroup>
        <FormGroup>
          <FormLabel>Date</FormLabel>
          <TransactionDatePicker
            name="date"
            dateFormat="DD/MM/YYYY"
            onChange={this.handleDateChange}
            selected={this.state.formData.date}
          />
        </FormGroup>
        <FormGroup>
          <FormLabel>Category</FormLabel>
          <FormSelect name="category" onChange={this.handleInputChange}>
            {options}
          </FormSelect>
        </FormGroup>
        <SubmitButton onClick={this.handleSubmit}> Submit </SubmitButton>
      </FormContainer>
    );
  }
}

export default TransactionForm;
