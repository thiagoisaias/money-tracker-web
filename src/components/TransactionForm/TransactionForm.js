import React, { Component } from "react";
import styled from "styled-components";
import { devices } from "../../utils/devices";

import Layout from "../Layout/Layout";

const FormContainer = styled.form`
  background-color: #fff;
  box-shadow: 0.5px 1px 1px 1px #ddd;

  > h2 {
    font-size: 1.5em;
    margin-bottom: 32px;
  }

  @media ${devices.small} {
    padding: 16px;
  }

  @media ${devices.mediumUp} {
    max-width: 900px;
    padding: 32px;
    margin: 32px auto;
    border-radius: 2px;
  }
`;

const FormGroup = styled.div`
  margin: 16px 0;
`;

const FormLabel = styled.label`
  display: block;
`;

const FormInput = styled.input`
  display: block;
  height: 30px;
  padding-left: 16px;
  border: 1px solid #e2e2e2;
  border-radius: 2px;
  margin: 8px 0;
  font-size: 14px;
  width: ${props => (props.bigger ? "300px" : null)};
`;

const FormSelect = styled.select`
  display: block;
  margin: 8px 0;
`;

const SubmitButton = styled.button`
  width: 125px;
  height: 40px;
  background-color: #8378f4;
  cursor: pointer;
  border-radius: 2px;
  color: #f2f2f2;
  font-size: 12px;

  &:hover {
    opacity: 0.9;
  }

  @media ${devices.small} {
    margin: 32px auto;
  }

  @media ${devices.mediumUp} {
    margin-top: 55px;
  }
`;

class TransactionForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      categoriesList: [],
      transactionData: {
        description: "",
        value: "",
        category: 0,
        transactionType: 0,
        date: Date.now()
      }
    };
  }

  handleInputChange = event => {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      transactionData: {
        ...this.state.transactionData,
        [name]: value
      }
    });
  };

  handleDateChange = event => {
    const selectedDate = event;
    this.setState({
      transactionData: {
        ...this.state.transactionData,
        date: selectedDate
      }
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    console.log("Send form data to API", this.state.transactionData);
  };

  render() {
    const options = this.state.categoriesList.map(category => {
      return (
        <option key={category.id} value={category.id}>
          {category.name}
        </option>
      );
    });

    return (
      <Layout>
        <FormContainer>
          <h2>Transaction</h2>
          <FormGroup>
            <FormLabel>Transaction Type</FormLabel>
            <FormSelect
              name="transactionType"
              onChange={this.handleInputChange}
            >
              <option value="0">Expense</option>
              <option value="1">Earning</option>
            </FormSelect>
          </FormGroup>
          <FormGroup>
            <FormLabel>Description</FormLabel>
            <FormInput
              bigger
              name="description"
              type="text"
              onChange={this.handleInputChange}
            />
          </FormGroup>
          <FormGroup>
            <FormLabel>Value</FormLabel>
            <FormInput
              name="value"
              type="number"
              onChange={this.handleInputChange}
            />
          </FormGroup>
          <FormGroup>
            <FormLabel>Date</FormLabel>
            <FormInput
              name="date"
              type="date"
              onChange={this.handleInputChange}
            />
          </FormGroup>
          <FormGroup>
            <FormLabel>Category</FormLabel>
            <FormSelect name="category" onChange={this.handleInputChange}>
              {options}
            </FormSelect>
          </FormGroup>
          <SubmitButton onClick={this.handleSubmit}> Submit </SubmitButton>
        </FormContainer>
      </Layout>
    );
  }
}

export default TransactionForm;
