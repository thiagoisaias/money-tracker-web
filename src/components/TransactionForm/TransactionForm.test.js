import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from 'enzyme-to-json';
import TransactionForm from "./TransactionForm";

describe("TransactionForm", () => {
  const component = shallow(<TransactionForm />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
