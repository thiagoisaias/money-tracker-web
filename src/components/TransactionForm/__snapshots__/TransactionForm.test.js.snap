// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionForm renders properly 1`] = `
<Layout>
  <styled.form>
    <h2>
      Transaction
    </h2>
    <styled.div>
      <styled.label>
        Transaction Type
      </styled.label>
      <styled.select
        name="transactionType"
        onChange={[Function]}
      >
        <option
          value="0"
        >
          Expense
        </option>
        <option
          value="1"
        >
          Earning
        </option>
      </styled.select>
    </styled.div>
    <styled.div>
      <styled.label>
        Description
      </styled.label>
      <styled.input
        bigger={true}
        name="description"
        onChange={[Function]}
        type="text"
      />
    </styled.div>
    <styled.div>
      <styled.label>
        Value
      </styled.label>
      <styled.input
        name="value"
        onChange={[Function]}
        type="number"
      />
    </styled.div>
    <styled.div>
      <styled.label>
        Date
      </styled.label>
      <styled.input
        name="date"
        onChange={[Function]}
        type="date"
      />
    </styled.div>
    <styled.div>
      <styled.label>
        Category
      </styled.label>
      <styled.select
        name="category"
        onChange={[Function]}
      />
    </styled.div>
    <styled.button
      onClick={[Function]}
    >
       Submit 
    </styled.button>
  </styled.form>
</Layout>
`;
